internal/auth/uaa_client.go

```diff
@@ -44,72 +44,110 @@ func NewUAAClient(
-		tokenCache:   make(map[string]uaaResponse),
+		uaa:        u,
+		httpClient: httpClient,
+		log:        log,
 	}
 }
 
-func (c *UAAClient) Read(token string) (Oauth2ClientContext, error) {
-	if token == "" {
-		return Oauth2ClientContext{}, errors.New("missing token")
+func (c *UAAClient) GetTokenKey() error {
+	req, err := http.NewRequest("GET", c.uaa.String(), nil)
+	if err != nil {
+		panic(fmt.Sprintf("failed to create request to UAA: %s", err))
 	}
 
-	uaaR, ok := c.tokenCache[token]
-	if ok && time.Now().Before(uaaR.ExpTime) {
-		return Oauth2ClientContext{
-			IsAdmin: c.hasDopplerScope(uaaR),
-			Token:   token,
-		}, nil
+	req.Header.Set("Content-Type", "application/json")
+
+	resp, err := c.httpClient.Do(req)
+
+	if err != nil {
+		return fmt.Errorf("failed to get token key from UAA: %s", err)
 	}
 
-	form := url.Values{
-		"token": {trimBearer(token)},
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("got an invalid status code talking to UAA %v", resp.Status)
 	}
+	defer resp.Body.Close()
 
-	req, err := http.NewRequest("POST", c.uaa.String(), strings.NewReader(form.Encode()))
+	tokenKey, err := decodeTokenKey(resp.Body)
 	if err != nil {
-		log.Printf("failed to create UAA request: %s", err)
-		return Oauth2ClientContext{}, err
+		return err
 	}
-	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
-	req.SetBasicAuth(c.client, c.clientSecret)
 
-	start := time.Now()
-	resp, err := c.httpClient.Do(req)
-	c.storeLatency(float64(time.Since(start)))
+	if tokenKey.Value == "" {
+		return fmt.Errorf("received an empty token key from UAA")
+	}
+
+	block, _ := pem.Decode([]byte(tokenKey.Value))
+	if block == nil {
+		return fmt.Errorf("failed to parse PEM block containing the public key")
+	}
+
+	publicKeyInterface, err := x509.ParsePKIXPublicKey(block.Bytes)
 	if err != nil {
-		log.Printf("UAA request failed: %s", err)
-		return Oauth2ClientContext{}, err
+		return fmt.Errorf("error parsing public key: %s", err)
 	}
 
-	defer func() {
-		io.Copy(ioutil.Discard, resp.Body)
-		resp.Body.Close()
+	sharedKey, isRSAPublicKey := publicKeyInterface.(*rsa.PublicKey)
+	if !isRSAPublicKey {
+		return fmt.Errorf("did not get a valid RSA key from UAA: %s", err)
+	}
+
+	c.sharedKey = sharedKey
+
+	return nil
+}
+
+func (c *UAAClient) StartPeriodicTokenKeyRefresh(duration time.Duration) {
+	go func() {
+		for {
+			time.Sleep(duration)
+			err := c.GetTokenKey()
+			if err != nil {
+				c.log.Printf("background token refresh failed: %s", err)
+			}
+		}
 	}()
+}
 
-	if resp.StatusCode != http.StatusOK {
-		return Oauth2ClientContext{}, fmt.Errorf("Non-200 status code received from UAA")
+func (c *UAAClient) Read(token string) (Oauth2ClientContext, error) {
+	if token == "" {
+		return Oauth2ClientContext{}, errors.New("missing token")
+	}
+
+	if c.sharedKey == nil {
+		return Oauth2ClientContext{}, errors.New("missing shared key from UAA")
+	}
+
+	payload, _, err := jose.Decode(trimBearer(token), c.sharedKey)
```

Comment:
How confident are we that this library won't accept `none` algoriths, or be tricked into doing HMAC over RSA etc?  Most JWT verifiers I've seen in Golang use https://github.com/dgrijalva/jwt-go which is pretty prescriptive in making you check the alg and supplied key. I haven't seen this one before...

e.g. in our impl we used as follows: https://github.com/govau/cf-common/blob/master/uaa/token_validator.go#L274